### PR TITLE
Add optional onColumnResize callback to IColumn

### LINKED ISSUE #6413: DetailsList: add callback to get column widths on resize
This feature will help us support overflowSets inside of columns. So rather than using a resizeGroup, or manually measuring column widths, if the DetailsList can report the widths it has determined for each column, we can manually adjust the items in the overflowSet.

### PR DESCRIPTION
**Pull request by KevinTCoughlin**
#### Pull request checklist

- [x] Addresses an existing issue: Fixes #6413
- [x] Include a change request file using `$ npm run change`

#### Description of changes

This feature is for each column to report its current width when resized so that consumers can manually adjust items within an `OverflowSet` (as an example) without using a `ResizeGroup` or manually measuring column widths. This feature is being added to help support `OverflowSet` within `IColumn`. It unblocks a Microsoft Flow scenario. 

**Perf Impact**

Below is my attempt at analyzing the perf impact before & after enabling `onColumnResize` for every column.

The example `DetailsList` used for this analysis is http://localhost:4322/#/examples/detailslist/advanced with `9` configured columns. The 'key' column was dragged from left to right-most edge and then back per "Performance" tab capture in Chrome.

**Before Change**

```shell
commitLifeCycles
Total Time
**8.57 ms**
Self Time
0.11 ms
Function
commitLifeCycles @ react-dom.development.js:14796
```

image

**After Change without onColumnResize set on any IColumn**

```shell
DetailsListBase._notifyColumnsResized
Total Time
10 μs
Self Time
0
Function
DetailsListBase._notifyColumnsResized @ DetailsList.base.tsx:405
```

image

**After Change with onColumnResize set on every IColumn**

```shell
commitLifeCycles
Total Time
**10.98 ms (+2.41 ms)**
Self Time
0.53 ms
Function
commitLifeCycles @ react-dom.development.js:10204

**DetailsListBase._notifyColumnsResized**
Total Time
**1.95 ms**
Self Time
0
Function
DetailsListBase._notifyColumnsResized @ DetailsList.base.tsx:405
```

image

**Summary**

Adding this change involves checking if each column has the optional `onColumnResize` prop set. The function is a no-op if none are set (existing behavior) so `_notifyColumnsResized` adds effectively `0` overhead. If the columns have this callback set, the time added is `~2.41ms` for `9` columns, roughly `.27ms` per column. This would grow linearly with the number of columns which have the property set.

Considering that the perf impact for consumers not using this optional prop (the majority right now) is effectively _zero_, I believe this is an OK feature to add that will not regress `DetailsList` performance.

#### Focus areas to test

- Add an `onColumnResize` callback to `IColumn` in a `DetailsList` example. One easy way to add it is via [buildColumns](https://github.com/OfficeDev/office-ui-fabric-react/blob/6505775ab433ec7e1fca27621be96dd6c026796d/packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.base.tsx#L1038).
- Drag a column or resize the viewport containing a `DetailsList` and the callback will be invoked per column which has it set.

**Note:** I'm currently figuring out a way to add unit tests for this prior to merge.

cc: @dliu120

### CODE REVIEW

**Comment by micahgodbolt** [comment]
Last question, does this fully solve the original customer's needs? Will this work on initial page load? This will get fired once the columns adjust to fit the screen size, right? Otherwise (if its not auto fitting) they will have defined the width anyway, right?

**Comment by KevinTCoughlin** [comment]
> 
> 
> Last question, does this fully solve the original customer's needs? Will this work on initial page load? This will get fired once the columns adjust to fit the screen size, right? Otherwise (if its not auto fitting) they will have defined the width anyway, right?

Great question! @dliu120 should confirm, but in our chat offline he believed this would fulfill their needs.

Re: when `_notifyColumnsResized` is invoked. It is called by `_adjustColumns` which is called:

1. On initial render via `componentWillReceiveProps`

    https://github.com/OfficeDev/office-ui-fabric-react/blob/6505775ab433ec7e1fca27621be96dd6c026796d/packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.base.tsx#L243

1. On column resize via `_onColumnResized`

    https://github.com/OfficeDev/office-ui-fabric-react/blob/6505775ab433ec7e1fca27621be96dd6c026796d/packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.base.tsx#L833

1. On column auto resize via `_onColumnAutoResized`

    https://github.com/OfficeDev/office-ui-fabric-react/blob/6505775ab433ec7e1fca27621be96dd6c026796d/packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.base.tsx#L870

1. On window `resize` via `withViewport` which invokes a `forceUpdate` on `List` => `DetailsList`

    https://github.com/OfficeDev/office-ui-fabric-react/blob/6505775ab433ec7e1fca27621be96dd6c026796d/packages/office-ui-fabric-react/src/utilities/decorators/withViewport.tsx#L105

On initial render it is invoked twice in my testing as `DetailsList` recalculates its layout so consumers should probably debounce it. Perhaps we can debounce it ourselves, but I'm worried of timing issues and reporting stale measurements.

**Comment by KevinTCoughlin** [comment]
Any tips on unit testing the above scenarios are appreciated. Drag will probably be the trickiest for me. Would like @dliu120 to confirm this addresses the need before wiring that up tho.

**Comment by micahgodbolt** [comment]
There are multiple ways to resize, correct? As long as they all result in the same function being called, i'd use the easiest one to reproduce to test this new functionality. i.e. if you can change the width of the parent container, and fire off this event, that's probably plenty test coverage.

**Comment by dliu120** [comment]
This should solve our needs! Thanks for adding this feature.

**Comment by KevinTCoughlin** [comment]
Sorry.. having some trouble testing `resize` here. Hoping to have a breakthru today.

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.test.tsx [hunk 1]
```diff
@@ -287,4 +287,29 @@ describe('DetailsList', () => {
...
+    const columns: IColumn[] = mockData(2, true);
+    columns[0].onColumnResize = jest.fn();
+    columns[1].onColumnResize = jest.fn();
+
+    // componentWillReceiveProps not executed on initial render in test
+    // so we need to force one via setProps and update.
+    const newProps = { columns };
+
+    detailsList.setProps(newProps);
+    detailsList.update();
```
Comment (KevinTCoughlin) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.test.tsx]: I was able to get the unit test working by using the above... cc: @natalieethell

**Comment by KevinTCoughlin** [comment]
@natalieethell asked for clarification between this `IColumn` callback and `DetailsList#onColumnResize` which I'll try to answer below.

## Summary

This new callback is mostly a convenience callback for devs, but it does have slightly different behavior. The differentiators which I believe justify the addition of this callback to `IColumn` are:

1. Each defined callback is invoked on `resize` whereas `DetailsList#onColumnResize` is not.
1. Each defined callback is invoked for convenience for the implementer whereas `DetailsList#onColumnResize` is only invoked for the _individual_ column that is resized.

## Differences in Behavior

`DetailsList#onColumnResize`:

- On column _drag_, reports the width after adjustment for the _individual_ column dragged
- On column _auto-adjust_, reports the width after adjustment for the _individual_ column dragged

`IColumn#onColumnResize`:

- On column _drag_, reports the width after adjustment for _each_ `IColumn` that has `onColumnResize` defined.
- On _auto-adjust_, reports the width after adjustment for _each_ `IColumn` that has `onColumnResize` defined.
- On viewport `resize`, reports the width after adjustment for _each_ `IColumn` that has `onColumnResize` defined.

## Test Code

```diff
--- a/packages/office-ui-fabric-react/src/components/DetailsList/examples/DetailsList.Basic.Example.tsx
+++ b/packages/office-ui-fabric-react/src/components/DetailsList/examples/DetailsList.Basic.Example.tsx
@@ -15,7 +15,10 @@ const _columns: IColumn[] = [
     minWidth: 100,
     maxWidth: 200,
     isResizable: true,
-    ariaLabel: 'Operations for name'
+    ariaLabel: 'Operations for name',
+    onColumnResize: (width: number) => {
+      console.log('Column 1', width);
+    }
   },
   {
     key: 'column2',
@@ -24,7 +27,10 @@ const _columns: IColumn[] = [
     minWidth: 100,
     maxWidth: 200,
     isResizable: true,
-    ariaLabel: 'Operations for value'
+    ariaLabel: 'Operations for value',
+    onColumnResize: (width: number) => {
+      console.log('Column 2', width);
+    }
   }
 ];
 
@@ -64,6 +70,10 @@ export class DetailsListBasicExample extends React.Component<
     };
   }
 
+  private onColumnResize(column?: IColumn, newWidth?: number, columnIndex?: number) {
+    console.log('DetailsList', 'onColumnResize', column, newWidth, columnIndex);
+  }
+
   public render(): JSX.Element {
     const { items, selectionDetails } = this.state;
 
@@ -90,6 +100,7 @@ export class DetailsListBasicExample extends React.Component<
             ariaLabelForSelectionColumn="Toggle selection"
             ariaLabelForSelectAllCheckbox="Toggle selection for all items"
             onItemInvoked={this._onItemInvoked}
+            onColumnResize={this.onColumnResize}
           />
         </MarqueeSelection>
       </div>
```

## Output

### On Initial Render

Note that only `IColumn#onColumnResize` is invoked.

```shell
11:34:00.407 Column 1 200
11:34:00.407 Column 2 200
```

### On Column Drag (Resize)

```shell
11:35:16.364 DetailsList onColumnResize {key: "column2", name: "Value", fieldName: "value", minWidth: 100, maxWidth: 200, …} 107 1
11:35:16.409 Column 1 200
11:35:16.410 Column 2 107
```

### On Column Auto Adjust

```shell
11:36:16.761 DetailsList onColumnResize {key: "column1", name: "Name", fieldName: "name", minWidth: 100, maxWidth: 200, …} 100 0
11:36:16.801 Column 1 100
11:36:16.801 Column 2 107
```

### On Viewport Resize

Note that only `IColumn#onColumnResize` is invoked.

```shell
11:46:36.120 Column 1 719
11:46:36.120 Column 2 813
```

**Comment by KevinTCoughlin** [comment]
Perhaps a comment block remark like the following would help as well:

```ts
  /**
   * If provided, will be executed when the column is resized with the column's current width.
   * Prefer this callback over {@link DetailsList.onColumnResize} if you require the `IColumn` to
   * report its width after every resize event. Consider debouncing the callback if resize events
   * occur frequently.
   */
```

**Pull request merged**